MNIST Classification using TensorFlow
======================================

In this notebook we will go through the steps to perform classification of handwritten digits in the MNIST data.

---
Loading Libraries and Dependencies
----------------------------------

In [2]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
%matplotlib inline

---
Part 1: Simple Problem - Single Layer Nueral Net
------------------------------------------------

In this we will build a single layer nueral net, which we will expand over time to create a deep nueral network.

We will use the data which comes as a part of the TensorFlow library.

In [3]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("MNIST_data", one_hot = True)

Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Extracting MNIST_data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [29]:
# Initialize all the weights and values
x = tf.placeholder(tf.float32, shape=[None, 784])
y = tf.placeholder(tf.float32, shape=[None, 10])

W = tf.Variable(tf.zeros([784, 10]))
b = tf.Variable(tf.zeros([10]))

# Define the model
y_model = tf.add(tf.matmul(x, W), b)

# Define the cost function
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(y_model, y))

# Define the optimizer
optimizer = tf.train.GradientDescentOptimizer(0.5).minimize(cost)
# Initialize all variables
init = tf.initialize_all_variables()

# Define the accuracy term
correct_prediction = tf.equal(tf.argmax(y,1), tf.argmax(y_model,1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

# Train the model
training_epochs = 1000
with tf.Session() as sess:
    sess.run(init)
    for i in range(training_epochs):
        batch = mnist.train.next_batch(100)
        sess.run(optimizer, feed_dict = {x:batch[0], y: batch[1]})
        if i %100 == 0:
            err = sess.run(accuracy, feed_dict = {x:batch[0], y: batch[1]})
            print "The train error of the model at iteration {:4d} is {:1.2f} %".format(i, err*100)

    acc = sess.run(accuracy, feed_dict = {x: mnist.test.images, y: mnist.test.labels})
    print "The test accuracy of our final model is {:1.2f} %".format(acc*100)


The train error of the model at iteration    0 is 48.00 %
The train error of the model at iteration  100 is 91.00 %
The train error of the model at iteration  200 is 92.00 %
The train error of the model at iteration  300 is 99.00 %
The train error of the model at iteration  400 is 93.00 %
The train error of the model at iteration  500 is 88.00 %
The train error of the model at iteration  600 is 95.00 %
The train error of the model at iteration  700 is 90.00 %
The train error of the model at iteration  800 is 93.00 %
The train error of the model at iteration  900 is 93.00 %
The test accuracy of our final model is 91.51 %


As we can see that our simple model achieved an accuracy of ~ 92% which is good but not great. To make our model peform better we will need to train a deeper nueral network. For this we will use what is called Convolution Nueral Nets.